xpp/astribank_tool.8
.TH "ASTRIBANK_TOOL" "8" "29 March 2009" "" ""

.SH NAME
astribank_tool \- Xorcom Astribank (xpp) control tool
.SH SYNOPSIS
.B astribank_tool [ options ] [ operation... ] \-D \fIdevice-path\fR

.B astribank_tool [\-h]

.SH DESCRIPTION
.B astribank_tool
is a tool to control the USB-level functionality of an Astribank.
The tool operates on a single Astribank at a time (given as parameter
to the \-D command line option).

.SH OPTIONS
.B \-D 
.I device-path
.RS
Required. The device to read from/write to. This is
\fIbus_num\fR/\fIdevice_num\fR, where \fIbus_num\fR and \fIdevice_num\fR
are the first two numbers in the output of lsusb(8) or dahdi_hardware(8).
On older versions of this tool you needed a complete path to the device,
which would be /dev/bus/usb/\fIbus_num\fR/\fIdevice_num\fR, or
/proc/bus/usb/\fIbus_num\fR/\fIdevice_num\fR.
.RE

.B \-p \fInum\fR
.RS
Set the TwinStar port number. Either 0 or 1.

(TODO: explain).
.RE

.B \-r \fItype\fR
.RS
Reset the Astribank and renumerate its USB connection to power on product ID.

Tyep can be: \fBhalf\fR or \fBfull\fR.

(TODO: explain those).
.RE

.B \-w 0|1
.RS
Enable (1) or disable (0) the TwinStar watchdog. When enabled, the
Astribank will jump to the second port if this system is "not working"
and the system on the second port is available.
.RE

.B \-Q
.RS
Query astribank properties via MPP protocol.
.RE

.B \-n
.RS
Renumerate the Astribank product number (e.g: from 1161 to 1162).
.RE

.B \-v
.RS
Increase verbosity. May be used multiple times.
.RE

.B \-d \fImask\fR
.RS
Set debug mask to \fImask\fR. Default is 0, 0xFF is "everything".
.RE

.B \-h
.RS
Displays usage message.
.RE

.SH SEE ALSO
fxload(8), lsusb(8), astribank_hexload(8)

.SH AUTHOR
This manual page was written by Tzafrir Cohen <tzafrir.cohen@xorcom.com> .
Permission is granted to copy, distribute and/or modify this document under
the terms of the GNU General Public License, Version 2 any 
later version published by the Free Software Foundation.

On Debian systems, the complete text of the GNU General Public
License can be found in /usr/share/common\-licenses/GPL.


ChangeLog
2011-10-21 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux 2.5.0.2 released.

2011-10-21 19:48 +0000 [r10253-10254]  Shaun Ruffell <sruffell@digium.com>

	* include/dahdi/kernel.h, drivers/dahdi/wctc4xxp/base.c: dahdi:
	  Move WARN_ON_ONCE from wctc4xxp driver to include/dahdi/kernel.h
	  I only generally test on RHEL 4 when testing against kernels
	  older than 2.6.18. Apparently OpenSUSE 10.1 runs with 2.6.16 and
	  doesn't have WARN_ON_ONCE backported. I took the patch Richard
	  Miller originally attached to the issue and moved it to
	  include/dahdi/kernel.h so it would be available for all the board
	  drivers in the future. Internal-Issue-ID: DAHLIN-260 Reported-by:
	  Richard Miller Signed-off-by: Shaun Ruffell <sruffell@digium.com>
	  Origin: http://svnview.digium.com/svn/dahdi?view=rev&rev=10252

	* drivers/dahdi/wcb4xxp/base.c: wcb4xxp: Do not show LASVEGAS2 as
	  echocan name if vpmsupport is set to 0 This fixes an issue where
	  "EC: LASVEGAS2" was displayed in /proc/dahdi/x for a B410P span
	  even though vpmsupport was disabled with the module parameter.
	  Internal-Issue-ID: DAHLIN-247 Signed-off-by: Shaun Ruffell
	  <sruffell@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10251

2011-10-12 16:32 +0000 [r10221]  Shaun Ruffell <sruffell@digium.com>

	* drivers/dahdi/wct4xxp/base.c: wct4xxp: Fix condition where
	  hardware echo canceler erroneously mutes DTMF. Commit r9750
	  "wct4xxp: Reduce the memory footprint of the hardware
	  echocanceler.", as part of reducing the non-pageable memory
	  required to support the VPMOCT064 and VPMOCT128, disabled caching
	  of some hardware echocan registers. This resulted in more
	  physical reads to the echo canceler. These new read transactions
	  exposed an existing issue where sometimes reads could be turned
	  into writes which put a channel into an unintended state
	  preventing Asterisk from detecting any DTMF. This issue is
	  resolved by ensuring that the write signal to the Octasic part is
	  explicitly cleared between when the address is presented on the
	  bus and when the read and chip select signals are asserted. The
	  cost is an increase in the average time to enable / disable echo
	  cancellation by about 5 us on one Intel Xeon X3220 test machine
	  (~250ns increase per read from the Octasic part and 20 reads to
	  enable / disable a channel). This commit resolves a behavioral
	  regression first introduced in 2.5.0 and 2.4.1 which could take
	  many calls before revealing itself. This change only affects
	  cards with a VPMOCT128 or VPMOCT064 installed. Signed-off-by:
	  Shaun Ruffell <sruffell@digium.com> Acked-by: Doug Bailey
	  <dbailey@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10220

2011-09-30 19:23 +0000 [r10219]  Shaun Ruffell <sruffell@digium.com>

	* drivers/dahdi/wctc4xxp/base.c: wctc4xxp: Allow G723 SID frames to
	  pass to the hardware decoder. The driver has, until now, dropped
	  G723 SID frames even though the firmware on the TC400/TCE400 can
	  handle them. Now let them on through. Reported-and-Tested-by:
	  Angel Carhuas <acarhuas@colinanet.com> Signed-off-by: Shaun
	  Ruffell <sruffell@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10218

2011-09-27 22:10 +0000 [r10211]  Shaun Ruffell <sruffell@digium.com>

	* drivers/dahdi/wctdm24xxp/base.c: wctdm24xxp: Set
	  dahdi_span.devicetype string in one place. Currently the
	  devicetype string was set both when the device was first
	  allocated and updated when an echocanceler was detected. For
	  simplicity, combine both these steps into a single function. This
	  change also replaces an improper use of strncat with strlcat.
	  Additionally, on the 2.5 branch, this change actually makes
	  r10206 "wctdm24xxp, wcte12xp: Advertise VPMOCT032 presence in
	  dahdi_span.devicetype", work the way it was originally intended.
	  That change was only functioning properly previously on trunk.
	  Signed-off-by: Shaun Ruffell <sruffell@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10210

2011-09-25 08:59 +0000 [r10208]  Tzafrir Cohen <tzafrir.cohen@xorcom.com>

	* drivers/dahdi/xpp/card_fxs.c: xpp: fxs: bugfix for 2fxs+6fxo
	  cards * Bug sympthoms: wrong FSK VMWI sent few seconds after
	  offhook. That was caused because the driver kept polling the
	  (physically unconnected) digital inputs. [note: a workaround for
	  drivers without this patch is to zero the
	  'xpd_fxs.poll_digital_inputs' parameter.] * Also, the
	  digital_inputs/digital_output masks were calculate using a
	  different condition. * Now we determine number of channels,
	  digital inputs and digital outputs in a single place and use this
	  info later to calculate the correct masks. * We poll only if
	  there are digital_inputs * We added a sanity check in
	  process_digital_inputs, so we get a notice if it's called on an
	  xpd without digital inputs (e.g: hypothetic firmware bug).
	  Signed-off-by: Tzafrir Cohen <tzafrir.cohen@xorcom.com> Acked-by:
	  Tzafrir Cohen <tzafrir.cohen@xorcom.com>

2011-09-23 20:23 +0000 [r10206-10207]  Shaun Ruffell <sruffell@digium.com>

	* drivers/dahdi/dahdi-base.c: dahdi: Check for master in
	  DAHDI_STARTUP / resolves MeetMe regression. There were a couple
	  of reports that MeetMe conferences were not working in 2.5.0.1
	  and that downgrading to 2.4.1.2 resolved the issue. This could
	  occur if there were no analog spans in a system, and all the
	  digital spans were out of alarm before DAHDI_STARTUP ioctl was
	  called by dahdi_cfg. If the spans were *not* out of alarm, they
	  would be marked master when the span changes it's alarm state.
	  This would result in a condition where no spans were marked as
	  the "master" and so the core timer was handling conferencing. The
	  core timer runs by default at 4ms and most board drivers run at
	  1ms intervals, but a channel currently only buffers up 2ms of

xpp/astribank_tool.c
/*
 * Written by Oron Peled <oron@actcom.co.il>
 * Copyright (C) 2008, Xorcom
 *
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <assert.h>
#include <getopt.h>
#include <sys/types.h>
#include <arpa/inet.h>
#include "astribank_usb.h"
#include "mpptalk.h"
#include <debug.h>
#include <xusb.h>

#define	DBG_MASK	0x80
/* if enabled, adds support for resetting pre-MPP USB firmware - if we 
 * failed opening a device and we were asked to reset it, try also the
 * old protocol.
 */
#define SUPPORT_OLD_RESET

static char	*progname;

static void usage()
{
	fprintf(stderr, "Usage: %s [options] -D {/proc/bus/usb|/dev/bus/usb}/<bus>/<dev> [operation...]\n", progname);
	fprintf(stderr, "\tOptions:\n");
	fprintf(stderr, "\t\t[-v]               # Increase verbosity\n");
	fprintf(stderr, "\t\t[-d mask]          # Debug mask (0xFF for everything)\n");
	fprintf(stderr, "\tOperations:\n");
	fprintf(stderr, "\t\t[-n]               # Renumerate device\n");
	fprintf(stderr, "\t\t[-r kind]          # Reset: kind = {half|full}\n");
	fprintf(stderr, "\t\t[-p port]          # TwinStar: USB port number [0, 1]\n");
	fprintf(stderr, "\t\t[-w (0|1)]         # TwinStar: Watchdog off or on guard\n");
	fprintf(stderr, "\t\t[-Q]               # Query device properties\n");
	exit(1);
}

static int reset_kind(const char *arg)
{
	static const struct {
		const char	*name;
		int		type_code;
	} reset_kinds[] = {
		{ "half",	0 },
		{ "full",	1 },
	};
	int	i;

	for(i = 0; i < sizeof(reset_kinds)/sizeof(reset_kinds[0]); i++) {
		if(strcasecmp(reset_kinds[i].name, arg) == 0)
			return reset_kinds[i].type_code;
	}
	ERR("Uknown reset kind '%s'\n", arg);
	return -1;
}


static int show_hardware(struct astribank_device *astribank)
{
	uint8_t	unit;
	uint8_t	card_status;
	uint8_t	card_type;
	int	ret;
	struct eeprom_table	eeprom_table;
	struct capabilities	capabilities;
	struct extrainfo	extrainfo;

	ret = mpp_caps_get(astribank, &eeprom_table, &capabilities, NULL);
	if(ret < 0)
		return ret;
	show_eeprom(&eeprom_table, stdout);
	show_astribank_status(astribank, stdout);
	if(astribank->eeprom_type == EEPROM_TYPE_LARGE) {
		show_capabilities(&capabilities, stdout);
		if(STATUS_FPGA_LOADED(astribank->status)) {
			for(unit = 0; unit < 5; unit++) {
				ret = mpps_card_info(astribank, unit, &card_type, &card_status);
				if(ret < 0)
					return ret;
				printf("CARD %d: type=%x.%x %s\n", unit,
						((card_type >> 4) & 0xF), (card_type & 0xF),
						((card_status & 0x1) ? "PIC" : "NOPIC"));
			}
		}
		ret = mpp_extrainfo_get(astribank, &extrainfo);
		if(ret < 0)
			return ret;
		show_extrainfo(&extrainfo, stdout);
		if(CAP_EXTRA_TWINSTAR(&capabilities)) {
			twinstar_show(astribank, stdout);
		}
	}
	return 0;
}

#ifdef SUPPORT_OLD_RESET
/* Try to reset a device using USB_FW.hex, up to Xorcom rev. 6885 */
int old_reset(const char* devpath)
{
	struct astribank_device *astribank;
	int ret;
	struct {
		uint8_t		op;
	} PACKED header = {0x20}; /* PT_RESET */
	char *buf = (char*) &header;

	/* Note that the function re-opens the connection to the Astribank
	 * as any reference to the previous connection was lost when mpp_open
	 * returned NULL as the astribank reference. */
	astribank = astribank_open(devpath, 1);
	if (!astribank) {
		DBG("Failed re-opening astribank device for old_reset\n");
		return -ENODEV;
	}
	ret = xusb_send(astribank->xusb, buf, 1, 5000);

	/* If we just had a reenumeration, we may get -ENODEV */
	if(ret < 0 && ret != -ENODEV)
			return ret;
	/* We don't astribank_close(), as it has likely been
	 * reenumerated by now. */
	return 0;
}	
#endif /* SUPPORT_OLD_RESET */

int main(int argc, char *argv[])
{
	char			*devpath = NULL;
	struct astribank_device *astribank;
	const char		options[] = "vd:D:nr:p:w:Q";
	int			opt_renumerate = 0;
	char			*opt_port = NULL;
	char			*opt_watchdog = NULL;
	char			*opt_reset = NULL;
	int			opt_query = 0;
	int			ret;

	progname = argv[0];
	while (1) {
		int	c;

		c = getopt (argc, argv, options);
		if (c == -1)
			break;

		switch (c) {
			case 'D':
				devpath = optarg;
				break;
			case 'n':
				opt_renumerate++;
				break;
			case 'p':
				opt_port = optarg;
				break;
			case 'w':
				opt_watchdog = optarg;
				break;
			case 'r':
				opt_reset = optarg;
				/*
				 * Sanity check so we can reject bad
				 * arguments before device access.
				 */
				if(reset_kind(opt_reset) < 0)
					usage();
				break;
			case 'Q':
				opt_query = 1;
				break;
			case 'v':
				verbose++;
				break;
			case 'd':
				debug_mask = strtoul(optarg, NULL, 0);
				break;
			case 'h':
			default:
				ERR("Unknown option '%c'\n", c);
				usage();
		}
	}
	if(!devpath) {
		ERR("Missing device path\n");
		usage();
	}
	DBG("Startup %s\n", devpath);
	if((astribank = mpp_init(devpath, 1)) == NULL) {
		ERR("Failed initializing MPP\n");
#ifdef SUPPORT_OLD_RESET
		DBG("opt_reset = %s\n", opt_reset);
		if (opt_reset) {
			DBG("Trying old reset method\n");
			if ((ret = old_reset(devpath)) != 0) {
				ERR("Old reset method failed as well: %d\n", ret);
			}
		}
#endif /* SUPPORT_OLD_RESET */

		return 1;
	}
	/*
	 * First process reset options. We want to be able
	 * to reset minimal USB firmwares even if they don't
	 * implement the full MPP protocol (e.g: EEPROM_BURN)
	 */
	if(opt_reset) {
		int	full_reset;

		if((full_reset = reset_kind(opt_reset)) < 0) {
			ERR("Bad reset kind '%s'\n", opt_reset);
			return 1;
		}
		DBG("Reseting (%s)\n", opt_reset);
		if((ret = mpp_reset(astribank, full_reset)) < 0) {
			ERR("%s Reseting astribank failed: %d\n",
				(full_reset) ? "Full" : "Half", ret);
		}
		goto out;
	}
	show_astribank_info(astribank);
	if(opt_query) {
		show_hardware(astribank);
	} else if(opt_renumerate) {
		DBG("Renumerate\n");
		if((ret = mpp_renumerate(astribank)) < 0) {
			ERR("Renumerating astribank failed: %d\n", ret);
		}
	} else if(opt_watchdog) {
		int	watchdogstate = strtoul(opt_watchdog, NULL, 0);

		DBG("TWINSTAR: Setting watchdog %s-guard\n",
			(watchdogstate) ? "on" : "off");
		if((ret = mpp_tws_setwatchdog(astribank, watchdogstate)) < 0) {
			ERR("Failed to set watchdog to %d\n", watchdogstate);
			return 1;
		}
	} else if(opt_port) {
		int	new_portnum = strtoul(opt_port, NULL, 0);
		int	tws_portnum = mpp_tws_portnum(astribank);
		char	*msg = (new_portnum == tws_portnum)
					? " Same same, never mind..."
					: "";

		DBG("TWINSTAR: Setting portnum to %d.%s\n", new_portnum, msg);
		if((ret = mpp_tws_setportnum(astribank, new_portnum)) < 0) {
			ERR("Failed to set USB portnum to %d\n", new_portnum);
			return 1;
		}
	}
out:
	mpp_exit(astribank);
	return 0;
}
